lib/web_ui/dev/steps/compile_tests_step.dart

```diff
@@ -0,0 +1,220 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io' as io;
+
+import 'package:path/path.dart' as pathlib;
+import 'package:pool/pool.dart';
+import 'package:web_test_utils/goldens.dart';
+
+import '../environment.dart';
+import '../utils.dart';
+import '../exceptions.dart';
+import '../watcher.dart';
+
+/// Compiles web tests and their dependencies.
+///
+/// Includes:
+///  * compile the test code itself
+///  * compile the page that hosts the tests
+///  * fetch the golden repo for screenshot comparison
+class CompileTestsStep implements PipelineStep {
+  CompileTestsStep({
+    this.skipGoldensRepoFetch = false,
+    this.testFiles,
+  });
+
+  final bool skipGoldensRepoFetch;
+  final List<FilePath>? testFiles;
+
+  @override
+  String get description => 'compile_tests';
+
+  @override
+  bool get isSafeToInterrupt => true;
+
+  @override
+  Future<void> interrupt() async {
+    await cleanup();
+  }
+
+  @override
+  Future<void> run() async {
+    if (!skipGoldensRepoFetch) {
+      await fetchGoldensRepo();
+    }
+    await buildHostPage();
+    await compileTests(testFiles ?? findAllTests());
+  }
+}
+
+/// Compiles the specified unit tests.
+Future<void> compileTests(List<FilePath> testFiles) async {
+  final Stopwatch stopwatch = Stopwatch()..start();
+
+  // Separate HTML targets from CanvasKit targets because the two use
+  // different dart2js options.
+  final List<FilePath> htmlTargets = <FilePath>[];
+  final List<FilePath> canvasKitTargets = <FilePath>[];
+  final String canvasKitTestDirectory =
+      pathlib.join(environment.webUiTestDir.path, 'canvaskit');
+  for (FilePath testFile in testFiles) {
+    if (pathlib.isWithin(canvasKitTestDirectory, testFile.absolute)) {
+      canvasKitTargets.add(testFile);
+    } else {
+      htmlTargets.add(testFile);
+    }
+  }
+
+  if (htmlTargets.isNotEmpty) {
+    await _compileTestsInParallel(targets: htmlTargets, forCanvasKit: false);
+  }
+
+  if (canvasKitTargets.isNotEmpty) {
+    await _compileTestsInParallel(targets: canvasKitTargets, forCanvasKit: true);
+  }
```

Comment:
Suggestion: in case both of these are true at the same time, should we compile them in parallel? Or do they have to be sequential?

```suggestion
  await Future.wait([
    if (htmlTargets.isNotEmpty)
      _compileTestsInParallel(targets: htmlTargets, forCanvasKit: false),
    if (canvasKitTargets.isNotEmpty)
      _compileTestsInParallel(targets: canvasKitTargets, forCanvasKit: true),
  ]);
```

Reply:
Done.